### PR TITLE
Add support for custom controls to SettingSourceAttribute

### PR DESCRIPTION
**Pull request by smoogipoo**
I'm not super set on having this support if there's disagreement as to whether it should be a thing, but I thought it'd be nice to have. Could be used for https://github.com/ppy/osu/pull/12568.

Follows the same model as `JsonConverter` - provide a `Type` to the attribute and it'll construct it. I haven't added it to the ctors, so you'd need to explicitly set the property via `[SettingSource(SettingControlType = typeof(X))]`.

Sample implementation:
```csharp
public class OsuModRandom : ModRandom, IApplicableMod
{
    public override string Description => "Random";

    [SettingSource("Seed", "", SettingControlType = typeof(OsuModRandomSettingsControl))]
    public BindableNumber<int> Seed { get; } = new BindableInt { Value = RNG.Next() };
}

public class OsuModRandomSettingsControl : SettingsItem<int>
{
    protected override Drawable CreateControl() => new ControlWithButton
    {
        RelativeSizeAxes = Axes.X,
        Margin = new MarginPadding { Top = 5 }
    };

    private class ControlWithButton : CompositeDrawable, IHasCurrentValue<int>
    {
        private readonly BindableWithCurrent<int> current = new BindableWithCurrent<int>();

        public Bindable<int> Current
        {
            get => current.Current;
            set => current.Current = value;
        }

        private readonly OsuTextBox textBox;

        public ControlWithButton()
        {
            AutoSizeAxes = Axes.Y;

            InternalChildren = new Drawable[]
            {
                new GridContainer
                {
                    RelativeSizeAxes = Axes.X,
                    AutoSizeAxes = Axes.Y,
                    ColumnDimensions = new[]
                    {
                        new Dimension(),
                        new Dimension(GridSizeMode.Absolute, 2),
                        new Dimension(GridSizeMode.Relative, 0.25f)
                    },
                    RowDimensions = new[]
                    {
                        new Dimension(GridSizeMode.AutoSize)
                    },
                    Content = new[]
                    {
                        new Drawable[]
                        {
                            textBox = new OsuNumberBox
                            {
                                RelativeSizeAxes = Axes.X,
                                CommitOnFocusLost = true
                            },
                            null,
                            new TriangleButton
                            {
                                RelativeSizeAxes = Axes.Both,
                                Height = 1,
                                Text = "Randomise!",
                                Action = randomiseSeed
                            }
                        }
                    }
                }
            };

            current.BindValueChanged(_ => updateDisplay(), true);
            textBox.Current.BindValueChanged(onTextBoxChanged);
            textBox.OnCommit += (_, __) => updateDisplay();
        }

        private void onTextBoxChanged(ValueChangedEvent<string> seed)
        {
            string newSeed = seed.NewValue;

            while (!string.IsNullOrEmpty(newSeed) && !int.TryParse(newSeed, out _))
                newSeed = newSeed[..^1];

            if (!int.TryParse(newSeed, out var intVal))
                intVal = 0;

            current.Value = intVal;
        }

        private void updateDisplay() => textBox.Text = current.Value.ToString();

        private void randomiseSeed() => current.Value = RNG.Next();
    }
}
```
image

### CODE REVIEW

File: osu.Game/Configuration/SettingSourceAttribute.cs [hunk 1]
```diff
@@ -31,7 +34,15 @@ public class SettingSourceAttribute : Attribute, IComparable<SettingSourceAttrib
 
         public int? OrderPosition { get; }
 
-        public SettingSourceAttribute(string label, string description = null)
+        /// <summary>
+        /// The type of the settings control which handles this setting source.
+        /// </summary>
+        /// <remarks>
+        /// Must be a type deriving <see cref="SettingsItem{T}"/> with a public constructor.
```
Comment (frenzibyte) [inline osu.Game/Configuration/SettingSourceAttribute.cs:41]: to complete it:
```suggestion
        /// Must be a type deriving <see cref="SettingsItem{T}"/> with a public parameterless constructor.
```

Comment (smoogipoo) [reply osu.Game/Configuration/SettingSourceAttribute.cs:41]: The constructor part is actually unnecessary.

Comment (frenzibyte) [reply osu.Game/Configuration/SettingSourceAttribute.cs]: oh? you'll need a public parameterless constructor to instantiate this type with `Activator.CreateInstance` though, unless you mean it's pretty obvious that it doesn't need to be mentioned?

Comment (smoogipoo) [reply osu.Game/Configuration/SettingSourceAttribute.cs]: Yeah you're right, I was thinking it's pretty obvious but I'll add it back in for safety...

**Review by frenzibyte — APPROVED** [review]
Sounds reasonable enough for me.

**Review by peppy — APPROVED** [review]
Seems fine. Definitely something we need sooner or later.

**Pull request merged**